tests/clone/clone.cpp

```diff
@@ -0,0 +1,550 @@
+
+        EXPECT_EQ(reference1, reference2);
+        EXPECT_EQ(prefix1, prefix2);
+        EXPECT_EQ(exponent1, exponent2);
+        EXPECT_EQ(multiplier1, multiplier2);
+        EXPECT_EQ(id1, id2);
+    }
+}
+
+void compareUnits(const libcellml::UnitsPtr &u1, const libcellml::UnitsPtr &u2, const libcellml::EntityPtr &expectedParent = nullptr)
+{
+    EXPECT_EQ(u1->id(), u2->id());
+    EXPECT_EQ(u1->isBaseUnit(), u2->isBaseUnit());
+    EXPECT_EQ(u1->isImport(), u2->isImport());
+    EXPECT_EQ(u1->importReference(), u2->importReference());
+    EXPECT_EQ(u1->name(), u2->name());
+    EXPECT_EQ(expectedParent, u2->parent());
+
+    compareUnit(u1, u2);
+}
+
+TEST(Clone, importSource)
+{
+    auto i = libcellml::ImportSource::create();
+
+    i->setId("import");
+    i->setUrl("this_is_a_url");
+
+    auto iClone = i->clone();
+
+    EXPECT_EQ(i->id(), iClone->id());
+    EXPECT_EQ(i->url(), iClone->url());
+    EXPECT_EQ(i->hasModel(), iClone->hasModel());
+}
+
+TEST(Clone, unitsBaseUnits)
+{
+    auto u = libcellml::Units::create();
+
+    u->setId("unique_id");
+    u->setName("units");
+    u->setImportReference("a_url");
+
+    auto uClone = u->clone();
+
+    compareUnits(u, uClone);
+}
+
+TEST(Clone, unitsImportedUnits)
+{
+    auto u = libcellml::Units::create();
+
+    auto import = libcellml::ImportSource::create();
+    import->setUrl("some-other-model.xml");
+
+    u->setId("unique_id");
+    u->setName("units");
+    u->setSourceUnits(import, "imported_units_name");
+
+    auto uClone = u->clone();
+
+    compareUnits(u, uClone);
+}
+
+TEST(Clone, unitsWithUnit)
+{
+    auto u = libcellml::Units::create();
+
+    u->setId("unique_id");
+    u->setName("units");
+    u->addUnit(libcellml::Units::StandardUnit::LUMEN);
+    u->addUnit("meter", "milli", -2.0);
+
+    auto uClone = u->clone();
+
+    compareUnits(u, uClone);
+}
+
+TEST(Clone, unitsImportedUnitsWithAdditionalUnit)
+{
+    auto u = libcellml::Units::create();
+
+    auto import = libcellml::ImportSource::create();
+    import->setUrl("some-other-model.xml");
+
+    u->setId("unique_id");
+    u->setName("units");
+    u->setSourceUnits(import, "imported_units_name");
+    u->addUnit(libcellml::Units::StandardUnit::LUMEN);
+    u->addUnit("a", "milli", -2.0);
+
+    auto uClone = u->clone();
+
+    compareUnits(u, uClone);
+}
+
+TEST(Clone, unitsInModel)
+{
+    auto m = libcellml::Model::create();
+    auto u = libcellml::Units::create();
+
+    u->setId("unique_id");
+    u->setName("units");
+
+    m->addUnits(u);
+
+    auto uClone = u->clone();
+
+    compareUnits(u, uClone);
+}
+
+TEST(Clone, variable)
+{
+    auto v = libcellml::Variable::create();
+    v->setUnits("lightning");
+    v->setInitialValue(-3.0);
+    v->setInterfaceType(libcellml::Variable::InterfaceType::PRIVATE);
+    v->setId("id_unique");
+    v->setName("variable");
+
+    auto vClone = v->clone();
+
+    EXPECT_EQ(nullptr, vClone->parent());
+    EXPECT_NE(v->units(), vClone->units());
```

Comment:
I take it that the units are not equal because of their parent being different? If so, it would be nice to explicitely check that? (Same for the other similar test below.)

Reply:
In fact, it would be nice to confirm in all our tests (e.g. `variableWithUnitsPtrSet `) that it is indeed the parent that is making the units different.

Reply:
Oops, I have just realised that I might have got it wrong. I was thinking of `compareUnits()` while here we are "simply" using `EXPECT_NE`, so it's probably checking the units address or something?

Reply:
This is comparing the address of the pointers to show that they are distinct and not pointing to the same resource.

Reply:
Ok, what I realised then.